dynamiqs/integrators/core/diffrax_integrator.py

```diff
@@ -225,3 +226,18 @@ def vector_field(t, y, _):  # noqa: ANN001, ANN202
             return tmp + tmp.dag()
 
         return dx.ODETerm(vector_field)
+
+
+class MCDiffraxIntegrator(DiffraxIntegrator, MCInterface):
+    """Integrator solving the Schrödinger equation with Diffrax."""
+
+    # subclasses should implement: diffrax_solver, discontinuity_ts
+
+    @property
+    def terms(self) -> dx.AbstractTerm:
+        def vector_field(t, y, _):  # noqa: ANN001, ANN202
+            L, H = self.L(t), self.H(t)
+            Hnh = sum_qarrays(-1j * H, *[-0.5 * _L.dag() @ _L for _L in L])
```

Comment:
Important comment: we should never call `_L.dag() @ _L` here, otherwise the complexity of the solver will shoot up to $O(n^3)$ in the dense case instead of the expected $O(n^2)$ for the Monte Carlo solver. This would really defeat its purpose.

Reply:
You don't actually need to compute the non-hermitian Hamiltonian, you only need to compute its action on a ket. So one way to go about it is to directly compute `sum_qarray(-1j * H @ y, *[-0.5 * _L.dag() @ (_L @ y) for _L in L])`. Whether that's the optimal way or writing it, I'm not sure, but it retains the $O(n^2)$ complexity.

Reply:
If possible, it would be great to benchmark this actually. I'm guessing which version has the shortest runtime will depend mostly on two factors:
 - Hilbert space size (yours should win in small sizes)
 - Sparsity (yours should win if all jump operators are sparse)

If this is true, we should probably add an option to choose between both versions (but we can start with just one for now to not overload the PR).

Reply:
Here's for instance something I wrote recently to benchmark different `mesolve` vector fields with and without vectorization of the Liouvillian:

```python
import jax
import jax.numpy as jnp
import dynamiqs as dq

# === Prepare objects ===
# parameters
n = 500
delta = 0.1
epsilon = 2
gamma = 1
nth = 0.8

# operators
a = dq.destroy(n)
H = delta * dq.dag(a) @ a + epsilon * (a + dq.dag(a))
L = [jnp.sqrt(gamma * (1 + nth)) * a, jnp.sqrt(gamma * nth) * dq.dag(a)]
y = dq.todm(dq.fock(n, 0))
liouv = dq.slindbladian(H, L)
yvec = dq.operator_to_vector(y)

# === Functions to benchmark ===
@jax.jit
def vf_dynamiqs(H, L, y):
    """Dynamiqs vector field."""
    L = dq.stack(L)
    Ld = L.dag()
    LdL = (Ld @ L).sum(0)
    tmp = (-1j * H - 0.5 * LdL) @ y + 0.5 * (L @ y @ Ld).sum(0)
    return tmp + tmp.dag()

@jax.jit
def vf_liouvillian(liouv, yvec):
    """Liouvillian vector field."""
    return liouv @ yvec

@jax.jit
def vf_liouvillian_dynamic(H, L, yvec):
    """Liouvillian vector field with dynamical Liouvillian computation."""
    liouv = dq.slindbladian(H, L)
    return liouv @ yvec

# === Run benchmarks ===
vf_dynamiqs(H, L, y).block_until_ready()
%timeit vf_dynamiqs(H, L, y).block_until_ready()
# n=50  / 167 µs ± 9.51 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=500 / 38.7 ms ± 10.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

vf_liouvillian(liouv, yvec).block_until_ready()
%timeit vf_liouvillian(liouv, yvec).block_until_ready()
# n=50  / 68.5 µs ± 4.41 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=500 / 2.32 ms ± 261 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

vf_liouvillian_dynamic(H, L, yvec).block_until_ready()
%timeit vf_liouvillian_dynamic(H, L, yvec).block_until_ready()
# n=50  / 134 µs ± 3.62 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=500 / 10.1 ms ± 808 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
Ahh I see! This is a very good point. Which would you prefer we go with for now for this PR before revisiting? I'm happy with either solution.

Reply:
From a quick benchmark, it seems my version is always faster actually. So I'd suggest going with this.

<details>
<summary>Here's the corresponding benchmark</summary>

```python
import jax
import jax.numpy as jnp
import dynamiqs as dq
from dynamiqs.qarrays.utils import sum_qarrays

dq.set_layout('dense')

# === Prepare objects ===
# parameters
n = 5000
delta = 0.1
epsilon = 2
gamma = 1
nth = 0.8

# operators
a = dq.destroy(n)
H = delta * dq.dag(a) @ a + epsilon * (a + dq.dag(a))
L = [jnp.sqrt(gamma * (1 + nth)) * a, jnp.sqrt(gamma * nth) * dq.dag(a)]
y = dq.fock(n, 0)

# === Functions to benchmark ===
@jax.jit
def vf1(H, L, y):
    Hnh = sum_qarrays(-1j * H, *[-0.5 * _L.dag() @ _L for _L in L])
    return Hnh @ y

@jax.jit
def vf2(H, L, y):
    return sum_qarrays(-1j * H @ y, *[-0.5 * _L.dag() @ (_L @ y) for _L in L])

# === Run benchmarks ===
vf1(H, L, y).block_until_ready()
%timeit vf1(H, L, y).block_until_ready()
# === sparse
# n=50   / 51.1 µs ± 5.74 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=500  / 50.6 µs ± 1.17 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=5000 / 59.2 µs ± 616 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# === dense
# n=50   / 86.9 µs ± 307 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=500  / 12.6 ms ± 640 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# n=5000 / 9.4 s ± 453 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

vf2(H, L, y).block_until_ready()
%timeit vf2(H, L, y).block_until_ready()
# === sparse
# n=50   / 49.1 µs ± 223 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=500  / 50.2 µs ± 855 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=5000 / 53.4 µs ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# === dense
# n=50   / 50 µs ± 3.21 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
# n=500  / 832 µs ± 62.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
# n=5000 / 208 ms ± 30.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
</details>

Reply:
Got it, thanks for running the benchmark!